### ISSUE TITLE
Core ML perf regression observed on non-genai models

### ISSUE THREAD

**Issue by guangy10** [pytorch/executorch#5564]
### 🐛 Describe the bug

We are running exercise on the [newly launched benchmarking infra](https://github.com/pytorch/executorch/tree/main/extension/benchmark) with the in-tree enabled models under `examples/models`. You can follow the instruction to extract and inspect the `benchmark_results.json`. (Connecting the results to a dashboard is still WIP) 

The preliminary data points collected from the benchmarking infra can be found here: https://github.com/pytorch/executorch/tree/main/extension/benchmark#preliminary-benchmark-results. It shows some previously enabled models, e.g. MobileNetV2/3, InceptionV3/4, are running expected slower with Core ML delegates.

The models are exported using the coreml export scripts ([code pointer](https://github.com/pytorch/executorch/blob/main/.ci/scripts/test_model.sh#L190)), and the app that loads and runs the exported model is located under [`extension/apple/benchmark`](https://github.com/pytorch/executorch/tree/main/extension/apple/Benchmark). 

### Versions

From latest executorch main

**Comment by cymbalrush** [comment]
Something weird is going on, the numbers are too high. Investigating.

**Comment by guangy10** [comment]
Here is the profiling of `mv3` running on the executor_runner binary:
```
...
│ 186 │ Execute            │ DELEGATE_CALL                                                       │  86.8285      │  86.8285      │  86.8285      │  86.8285      │  86.8285      │  86.8285      │ ['aten.convolution.default', 'aten._native_batch_norm_legit_no_training.default' ... 'aten.addmm.default'] (317 total) │ False             │ CoreMLBackend           │ nan               │           nan           │ nan                       │ nan                        │
├─────┼────────────────────┼─────────────────────────────────────────────────────────────────────┼───────────────┼───────────────┼───────────────┼───────────────┼───────────────┼───────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼───────────────────┼─────────────────────────┼───────────────────┼─────────────────────────┼───────────────────────────┼────────────────────────────┤
│ 187 │ Execute            │ Method::execute                                                     │  86.8298      │  86.8298      │  86.8298      │  86.8298      │  86.8298      │  86.8298      │ []                                                                                                                     │ False             │                         │ nan               │           nan           │ nan                       │ nan                        │
╘═════╧════════════════════╧═════════════════════════════════════════════════════════════════════╧═══════════════╧═══════════════╧═══════════════╧═══════════════╧═══════════════╧═══════════════╧════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╧═══════════════════╧═════════════════════════╧═══════════════════╧═════════════════════════╧═══════════════════════════╧════════════════════════════╛
```
The avg of `Method::execute` is 86.829ms, far different from the numbers get from the benchmark app. @shoumikhin Is it same as how test_forward is measured in the app?

**Comment by YifanShenSZ** [comment]
Do you see any graph breaks?

**Comment by cymbalrush** [comment]
@YifanShenSZ  There are no graph breaks.

@guangy10 the actual number is close to 1 ms. The `86.829` ms you are seeing is because the delegate execution time is the total time taken by the execute call. When profiling, there is some cost associated with getting the profile data and writing it. 

@Olivia-liu  Is there a way for a delegate to log the time directly instead of relying on the execute  duration? 

@shoumikhin are the numbers still low?

**Comment by Olivia-liu** [comment]
@cymbalrush @guangy10 

Not exactly log the time directly, but should be helpful to reduce most overhead:

I found `EventTracerProfilingLevel::kProfileMethodOnly` https://github.com/pytorch/executorch/blob/1052e3b17ee0750c89ae78dd60c61dd6b54ebb89/runtime/core/event_tracer.h#L77

So basically it's an known issue that the overhead for profiling ops can be high. And one can use this option to disable per op profiling and essentially only profile the important events such as inference execution time. 

In this case since @guangy10 is using the executor_runner, in the runner implementation, find the line `ETDumpGen etdump_gen;` and add `etdump_gen. set_event_tracer_profiling_level(EventTracerProfilingLevel::kProfileMethodOnly)`

**Comment by cymbalrush** [comment]
Thanks @Olivia-liu! for the pointer but is this something that the tooling can address? otherwise this will keep on bugging us :)

**Comment by Olivia-liu** [comment]
@cymbalrush Yes, will do. Is the request that while per op profiling is enabled in the delegate calls, the duration of the delegate call itself is measured or calculated in such a way that the profiling overhead is removed?

**Comment by tarun292** [comment]
Hmm @cymbalrush @guangy10 something seems off here. Can you share the etdump here? Profiling has overhead but it shouldn't be that much.

**Comment by guangy10** [comment]
The root cause was how the numbers are reported. We have re-calibrated the new ios benchmark app and verified the perf numbers on-device. The number are updated here: https://github.com/pytorch/executorch/tree/main/extension/benchmark#preliminary-benchmark-results.

**Comment by cccclai** [comment]
> Hmm @cymbalrush @guangy10 something seems off here. Can you share the etdump here? Profiling has overhead but it shouldn't be that much.

For the more prominent overhead (small models like resnet18), it seems like the devtools delegate execute time is 0.98 ms, while the trace shows 0.358 ms. Any chance you know the overhead of the coreml layer? [Since the delegate call is pretty much one line call](https://github.com/pytorch/executorch/blob/f0112a2c63b16c6d69f487731691c22b5b8bdede/runtime/executor/method.cpp#L126-L127), the overhead ideally shouldn't be that much. 

The issue reported as the title, like the 86 ms overhead, might be the benchmark infra issue. We may file a different issue to track them seperately for the delegate profiling overhead.

**Comment by cccclai** [comment]
Should we close this one?

**Comment by kimishpatel** [comment]
Thanks @cccclai for flagging to close this. I am gonna close this for now
